### PR TITLE
refs #2871 fixed a bug handling duplicate soapAction values

### PR DESCRIPTION
**Pull request by davidnich**

### CODE REVIEW

**Comment by davidnich** [comment]
@sejvlond please do not merge this yet - adding other emergency fixes

**Comment by davidnich** [comment]
@sejvlond now it's OK

File: qlib/WSDL.qm [hunk 1]
```diff
@@ -4429,7 +4434,8 @@ public class WSDL::WebService inherits WSDL::XsdBase {
           - \c "binding": the binding of the port
      */
     list<hash<WSDL::ServiceInfo>> listServices() {
-        return cast<list<hash<WSDL::ServiceInfo>>>(services.values());
+        return map cast<hash<WSDL::ServiceInfo>>($1), services.values();
+        #return cast<list<hash<WSDL::ServiceInfo>>>(services.values());
```
Comment (sejvlond) [inline qlib/WSDL.qm]: remove the commented code

File: qlib/SoapClient.qm [hunk 2]
```diff
@@ -360,10 +360,12 @@ public namespace SoapClient {
...
             @throw HTTP-CLIENT-RECEIVE-ERROR this exception is thrown when the SOAP server returns an HTTP error code; if a SOAP fault is returned, then it is deserialized and returned in the \a arg key of the exception hash
 
-            @note this method can throw any exception that @ref Qore::HTTPClient::send() "HTTPClient::send()" can throw as well as any XML parsing errors thrown by @ref Qore::XML::parse_xml() "parse_xml()"
+            @note
+            - this method can throw any exception that @ref Qore::HTTPClient::send() "HTTPClient::send()" can throw as well as any XML parsing errors thrown by @ref Qore::XML::parse_xml() "parse_xml()"
+            - it's recommended to use callOperation() instead, which was a cleaner and more extensible API
         */
-        any call(string operation, any args, *hash header, *hash nsh, *string binding) {
-            return makeCallIntern(NOTHING, operation, args, ("soap_header": header, "http_header": nsh, "binding": binding));
+        auto call(string operation, auto args, *hash<auto> header, *hash<auto> nsh, *string binding) {
```
Comment (sejvlond) [inline qlib/SoapClient.qm]: can you maybe rename the `header` and `nsh` to the `soap_header` and `http_header` all over the module?

**Review by sejvlond — CHANGES_REQUESTED** [review]

File: qlib/WSDL.qm (hunk 1 above)
Comment (davidnich) [reply qlib/WSDL.qm]: actually let me fix https://github.com/qorelanguage/qore/issues/2877 first and then do this right

File: qlib/SoapClient.qm (hunk 2 above)
Comment (davidnich) [reply qlib/SoapClient.qm]: ok will do

Comment (davidnich) [reply qlib/SoapClient.qm]: done

File: qlib/WSDL.qm (hunk 1 above)
Comment (davidnich) [reply qlib/WSDL.qm]: done

**Comment by davidnich** [comment]
@sejvlond all issues should be addressed; please check

**Review by sejvlond — APPROVED** [review]

**Pull request merged**